### ISSUE TITLE
Issue #4: Analytical Ruin Probability

### ISSUE THREAD

**Issue by Fenris91** [Fenris91/mlfinlab_test-3#7]
Labels: enhancement, priority-4, risk
Context
Our Monte Carlo simulation estimates 8.1% ruin probability for XGB-full. We need the closed-form solution to cross-validate and to quickly evaluate how ruin probability changes with different Kelly fractions.
Requirements
Create src/risk/ruin_probability.py with:
pythondef ruin_probability_discrete(
    win_rate: float,
    payoff_ratio: float,
    bet_fraction: float,
    initial_bankroll: float = 10000.0,
    ruin_threshold: float = 0.0,
) -> float:
    """
    Classical gambler's ruin for asymmetric payoffs.
    Uses recursive formula for exact solution when possible,
    falls back to numerical approximation.
    """

def ruin_probability_continuous(
    sharpe: float,
    bet_fraction: float,
    bankroll_to_ruin_ratio: float,
) -> float:
    """
    Continuous approximation: P(ruin) ≈ exp(-2 * μ * D / σ²)
    where D = distance to ruin, μ = drift, σ = volatility.
    """

def ruin_vs_kelly_curve(
    win_rate: float,
    payoff_ratio: float,
    fractions: np.ndarray | None = None,
    n_points: int = 50,
) -> dict:
    """
    Compute ruin probability across a range of Kelly fractions.
    Returns: {'fractions': array, 'ruin_probs': array, 'full_kelly': float}
    Useful for visualizing the risk/reward tradeoff.
    """

def max_fraction_for_target_ruin(
    win_rate: float,
    payoff_ratio: float,
    target_ruin: float = 0.05,
) -> float:
    """
    Find the maximum bet fraction that keeps P(ruin) below target.
    Binary search over ruin_probability_discrete.
    """
Tests
Create tests/test_ruin.py:

test_fair_coin_ruin: win=0.5, payoff=1.0 → P(ruin) approaches 1.0 for any fraction
test_strong_edge_low_ruin: win=0.6, payoff=2.0 → P(ruin) < 0.01 at half-Kelly
test_our_strategy: win=0.395, payoff=2.96 → compare with MC 8.1%
test_continuous_vs_discrete: both methods agree within tolerance
test_ruin_increases_with_fraction: monotonically increasing
test_kelly_curve_shape: ruin near 0 at small fractions, near 1 at 2x Kelly
test_max_fraction_5pct: target_ruin=0.05 → returns fraction below full Kelly

Acceptance Criteria

 All tests pass
 Analytical result within 2% of MC 8.1% for our strategy parameters
 Only numpy + scipy dependencies
 Handles edge cases (negative edge, zero bankroll, payoff_ratio=1)

Issue #5: Combinatorial Purged Cross-Validation (CPCV)
Labels: enhancement, priority-5, validation
Context
Our walk-forward uses 42 sequential folds. Bootstrap resampling of these folds gives confidence intervals, but CPCV generates all possible train/test path combinations with proper temporal purging, producing tighter estimates. This extends mlfinlab/cross_validation/.
Requirements
Create src/cross_validation/cpcv.py with:
pythondef cpcv_paths(
    n_groups: int,
    n_test_groups: int,
    purge_groups: int = 1,
    embargo_groups: int = 0,
) -> list[tuple[list[int], list[int]]]:
    """
    Generate all combinatorial train/test splits.
    Each path uses n_test_groups for testing, rest for training.
    Purge: remove purge_groups adjacent to test boundary.
    Embargo: skip embargo_groups after test set.
    Returns: list of (train_indices, test_indices) tuples.
    """

def cpcv_backtest_paths(
    n_groups: int,
    n_test_groups: int,
    purge_groups: int = 1,
) -> list[list[int]]:
    """
    Generate all possible backtest paths through the data.
    Each path visits every time period exactly once as a test set.
    Returns: list of paths, where each path is a sequence of group indices.
    """

def cpcv_statistics(
    fold_returns: np.ndarray,  # (n_groups,) per-fold returns
    n_test_groups: int,
    purge_groups: int = 1,
) -> dict:
    """
    Compute path-wise statistics from pre-computed fold returns.
    Returns: {
        'n_paths': int,
        'path_sharpes': np.ndarray,
        'mean_sharpe': float,
        'median_sharpe': float,
        'sharpe_95_ci': tuple[float, float],
        'p_sharpe_positive': float,
        'p_sharpe_above_benchmark': float,
    }
    """

def probability_of_backtest_overfitting(
    fold_returns: np.ndarray,
    n_test_groups: int,
    purge_groups: int = 1,
) -> float:
    """
    PBO: fraction of CPCV paths where in-sample optimal
    strategy underperforms out-of-sample.
    López de Prado (2018), Chapter 11.
    """
Reference

López de Prado, M. (2018). "Advances in Financial Machine Learning", Chapter 12
Bailey et al. (2017). "The Probability of Backtest Overfitting"
Key insight: with N groups choosing k test groups, there are C(N,k) paths

Tests
Create tests/test_cpcv.py:

test_path_count: n=6, k=2 → C(6,2) = 15 paths
test_purge_reduces_train: purge=1 removes groups adjacent to test
test_embargo_skips_after_test: embargo groups excluded from training
test_all_groups_tested: every group appears in test set across paths
test_no_train_test_overlap: no group appears in both train and test
test_statistics_known_returns: synthetic fold returns → verify CI calculation
test_pbo_random_returns: random returns → PBO near 0.5 (no skill)
test_pbo_perfect_returns: monotonically increasing → PBO near 0.0

Acceptance Criteria

 All tests pass
 Compatible with pre-computed fold returns (doesn't need to re-run models)
 Only numpy + scipy + itertools dependencies
 Handles edge cases (n_test_groups > n_groups/2, purge larger than available)
 Path enumeration is efficient (C(42,2) = 861 paths is manageable)
